Egor/zed extension

egol opened the pull request:
Fixes #932

### Changes:
- Embed playground server in the language server (LSP).
  - Allows support for any editor that supports LSP protocol.
- Modify `language_server/server.rs` to launch the server on first available localhost starting at port `3030`. If `3030` is taken will try `3031` etc.
  - Sends events on every LSP `did_change`
  - Selects function on function ctrl + hover
  - Code actions for opening playground in browser
- Create a new zed extension under `engine/zed` which packages the new embedded playground via LSP
  - **Embedding structure:** Wasm language runtime -> fiddle frontend -> baml-cli (LSP) -> zed-extension -> Wasm extension package. 
  - Syntax highlighting partially implemented using the existing tree-sitter implementation
- Added connection status for playground frontend on connection lost

### Known Issues:
- Syntax highlighting is not fully implemented:
  - Jinja highlighting does not work
  - Incorrect field and type highlighting

### Notes:
- May be useful to batch updates from LSP to not spam frontend
- Entire frontend is now embedded in LSP increasing file size
- Local baml version handling is currently not implemented
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Embed playground server in language server for LSP support and add Zed extension for packaging, with partial syntax highlighting and connection status indicator.
> 
>   - **Behavior**:
>     - Embed playground server in `language_server` to support any LSP-compatible editor.
>     - Launches on first available localhost port starting at `3030`.
>     - Sends events on LSP `did_change`, supports function selection on hover, and code actions for opening playground.
>   - **Zed Extension**:
>     - New extension under `engine/zed` packages embedded playground via LSP.
>     - Embedding structure: Wasm language runtime -> fiddle frontend -> baml-cli (LSP) -> zed-extension -> Wasm extension package.
>     - Partial syntax highlighting using tree-sitter.
>   - **Playground Frontend**:
>     - Added connection status indicator for connection loss.
>   - **Known Issues**:
>     - Incomplete syntax highlighting: Jinja and incorrect field/type highlighting.
>     - Local baml version handling not implemented.
>   - **Misc**:
>     - Added `build.sh` script in `language_server/scripts` for building web-panel.
>     - Updated `Cargo.toml` and `Cargo.lock` for new dependencies and configurations.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup> for 7cc5efbac3cf6e3ef7761bf2a95c63b65b038d19. You can [customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

egol committed:
basic hosting of frontend and websocket communication

egol committed:
rust talking to frontend

egol committed:
cleaned up code

egol committed:
add basic file watching for saved files

egol committed:
improve file watching to watch a baml_src directory

egol committed:
added init which grabs all baml files at baml_src

egol committed:
basic implementation for playground via cli

egol committed:
correctly update removed .baml files

egol committed:
change embedd to pull from web-panel build file directly

egol committed:
basic LSP playground embed

egol committed:
add reciever to keep broadcast channel alive

egol committed:
web update on save working

egol committed:
did_change test

egol committed:
working unsaved update passthrough from lsp to playground

egol committed:
Connection error for webview

egol committed:
init zed extension support

egol committed:
added on hover function switching to LSP

egol committed:
disable websocket in vscode

egol committed:
automatically find a new port if selected one is being used

egol committed:
change git path for extension and set to release mode

egol committed:
revert changes to cli

egol committed:
Merge branch 'canary' into Egor/zed-extension

aaronvg commented on typescript/playground-common/src/baml_wasm_web/EventListener.tsx:
```diff
@@ -112,18 +138,47 @@ export const EventListener: React.FC<{ children: React.ReactNode }> = ({ childre
   }, [selectedFunc])
   console.log('selectedFunc', selectedFunc)
 
+  useEffect(() => {
```
if we push this do we lose functionality in VSCode to send messages to the webview playground? VSCode doesn't yet use this bundled playground does it?

aaronvg replied:
basically my comment is -- do we have backwards compatibility with the curent vscode extension

egol replied:
It does not currently do anything to the functionality of the vscode extension. It works in parallel, which while redundant does not break anything.

aaronvg commented on engine/cli/Cargo.toml:
```diff
@@ -77,6 +77,7 @@ web-time.workspace = true
 static_assertions.workspace = true
 mime_guess = "2.0.4"
 mime = "0.3.17"
+include_dir = "0.7"
```
remove this

aaronvg commented on engine/language_server/src/playground/playground.rs:
```diff
@@ -0,0 +1,26 @@
+
+    pub async fn run(self, port: u16) -> Result<()> {
+        let routes = create_routes(self.state, self.session);
+        warp::serve(routes).run(([127, 0, 0, 1], port)).await;
```
what happens when this fails

egol replied:
changed to use `try_bind`

aaronvg commented on engine/language_server/src/playground/playground_server.rs:
```diff
@@ -0,0 +1,172 @@
+    }
+}
+
+pub async fn client_connection(
```
is this more of a start_client_connection? Naming could be improved

egol replied:
Agreed

aaronvg commented on engine/language_server/src/playground/playground_server.rs:
```diff
@@ -0,0 +1,172 @@
+    });
+}
+
+pub fn create_routes(
```
better name would be create_server_routes, and add a comment that we only add a route "/" that serves the static files, and another one to listen to websocket events.

aaronvg commented on engine/language_server/src/server.rs:
```diff
@@ -369,4 +379,45 @@ impl Server {
+                        break;
+                    } else {
+                        // Port is already in use, try next port
+                        playground_port += 1;
```
fyi you'll need to send the port to the extension client later so they can point to the right url

egol replied:
Fixed via [commit](https://github.com/BoundaryML/baml/pull/2044/commits/4f638f29c5f3527c1bff27b84d1c3276c094fb4d)

egol committed:
force web-view to be built before language server

egol committed:
various fixes + code actions for zed

egol committed:
disable on-hover fn selection and add go to def selection. Fix connection status as well

egol committed:
build modification

egol committed:
build script adjustments

egol committed:
more build script adjustments

egol committed:
build script runs after macro in rust, remove macro for embed

egol committed:
try changing workflow file order

egol committed:
more workflow modification

egol committed:
fix missing wasm-bindgen when building language-server

egol committed:
download specfic wasm-bindgen version in install script

egol committed:
fix test for lsp and git diff check

egol committed:
remove build.rs script and replace with install.sh, fix connection status

egol committed:
fix order of install script

egol committed:
add comments and rename script

hellovai committed:
Greg/go (#2045)

Co-authored-by: Greg Hale <imalsogreg@gmail.com>
Co-authored-by: Rahul Tiwari <jprrahultiwari@gmail.com>

aaronvg committed:
Boundary studio v2 model changes (#2048)

- Boundary studiov2 model changes
<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Enhance BAML runtime tracing and logging, add HTTP request/response
handling, and improve BAML source hashing performance.
> 
>   - **Tracing and Logging Enhancements**:
> - Add `FunctionType` and `FunctionStart` to `tracing::events` for
better function call tracing.
> - Implement `IntoRpcEvent` for converting trace data to RPC events in
`rpc_converters`.
> - Add `make_trace_event_for_response()` in
`llm_response_to_log_event.rs` to create trace events from LLM
responses.
>   - **HTTP Handling**:
> - Add `HTTPRequest` and `HTTPResponse` handling in Python, Ruby, and
TypeScript clients.
> - Implement header redaction for sensitive information in
`tracing/events.rs`.
>   - **Performance Improvements**:
> - Add `baml_src_hash()` in `AstSignatureWrapper` for deterministic
hashing of BAML source code.
> - Add performance tests for BAML source hashing in `publisher/mod.rs`.
>   - **Miscellaneous**:
>     - Comment out unused code in `upload-traces-main.rs`.
>     - Update `Cargo.lock` with new dependencies and versions.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for fdc64e06c09a906504aa7b7e4de0b85c2a3475bf. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

---------

Co-authored-by: Sam Lijin <sam@boundaryml.com>
Co-authored-by: Chris Watts <chris@boundaryml.com>

imalsogreg committed:
Add linter warnings for experimental expr features (#2053)

<!-- ELLIPSIS_HIDDEN -->


> [!IMPORTANT]
> Add linter warnings for experimental workflow functions and variable
assignments in `expr_fns.rs`.
> 
>   - **Warnings**:
> - Adds `DatamodelWarning` for experimental workflow functions in
`validate_expr_fns()`.
> - Adds `DatamodelWarning` for experimental variable assignments in
`validate_expr_fns()`.
>   - **Imports**:
>     - Adds `DatamodelWarning` to imports in `expr_fns.rs`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 10525dc195fb29ea769b119a93a4131bd07f7556. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

ba11b0y committed:
run cargo fix (#2059)

<!-- ELLIPSIS_HIDDEN -->



> [!IMPORTANT]
> Run `cargo fix` to remove unused imports in `cffi_generated.rs` and
`lib.rs`.
> 
>   - **Imports**:
> - Remove unused imports `core::mem` and `core::cmp::Ordering` from
`cffi_generated.rs`.
> - Remove unused import
`baml_runtime::tracingv2::storage::storage::Usage` from `lib.rs`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 9574b86c6887e59d425002ed57a4efe8eb5606c1. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

ba11b0y committed:
fix: pass in env vars for cli tests (#2060)

A bug was introduced in #1949 which was not injecting env vars for
`baml-cli test`
This PR fixes that by passing in the environment variables.

Fixes #2051 

<img width="880" alt="image"
src="https://github.com/user-attachments/assets/6df90bcc-4ef7-4d7c-9d68-e938957ef0e3"
/>

<!-- ELLIPSIS_HIDDEN -->


----

> [!IMPORTANT]
> Fixes bug in `baml-cli test` by passing environment variables to
`BamlRuntime` and `TestExecutor`.
> 
>   - **Behavior**:
> - Fixes bug in `baml-cli test` by passing environment variables to
`BamlRuntime::from_directory` and `TestExecutor::cli_run_tests`.
> - Collects environment variables into `HashMap` and passes them to
`BamlRuntime` and `TestExecutor`.
>   - **Functions**:
> - Updates `BamlRuntime::from_directory` in `lib.rs` to accept
`env_vars`.
> - Updates `TestExecutor::cli_run_tests` in `test_executor/mod.rs` to
accept `env_vars`.
>   - **Misc**:
>     - Removes redundant logging of environment variables in `lib.rs`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for db380e267a581febc9a21aab161c2c5b2d741d54. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

antoniosarosi committed:
Rename `SchemaAst` to `Ast` and `schema-ast` crate to `ast` (#2052)

<!-- ELLIPSIS_HIDDEN -->


> [!IMPORTANT]
> Renames `SchemaAst` to `Ast` and `schema-ast` crate to `ast`, updating
imports and function names accordingly.
> 
>   - **Renames**:
>     - `SchemaAst` to `Ast` in `ast.rs` and 8 other files.
>     - `schema-ast` crate to `ast` in `Cargo.toml` and 2 other files.
>     - `parse_schema` to `parse` in `parse.rs` and 2 other files.
>   - **Imports**:
> - Update import paths from `internal-baml-schema-ast` to
`internal-baml-ast` in `mod.rs` and 10 other files.
>   - **Misc**:
>     - Update `Cargo.lock` to reflect new crate name.
>     - Update `format.rs` to use `internal_baml_ast` for formatting.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 93d7babe987fbf6de9c27c998914146b5aedd85a. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

ba11b0y committed:
fix: blocking within a runtime (#2061)

While running `baml-cli test` for multiple tests, we found out that we
were blocking on a tokio runtime to create a new runtime. When creating
a new runtime, we would already have flushed out events from the
previous one, we would not need to block here.
<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Removes blocking calls in `start_publisher()` in `publisher.rs` to
improve runtime creation performance.
> 
>   - **Behavior**:
> - Removes blocking calls to `flush()` in `start_publisher()` in
`publisher.rs` for non-WASM32 architectures.
> - Ensures events are flushed before creating a new runtime without
blocking.
>   - **Misc**:
> - Removes conditional compilation code for blocking on `flush()` in
`start_publisher()`.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 1aa9facad7dc2ae583f752d90115528ce372600b. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

antoniosarosi committed:
Add `cargo fmt --check` to CI (#2062)

<!-- ELLIPSIS_HIDDEN -->


> [!IMPORTANT]
> Add `cargo fmt --check` to CI to enforce Rust code formatting and
improve import organization across several files.
> 
>   - **CI/CD**:
> - Add `cargo fmt --check` to `primary.yml` to enforce Rust code
formatting in CI.
>   - **Code Organization**:
>     - Remove unused imports in `entry.rs` and `mod.rs`.
> - Reorganize imports in `runtime_interface.rs`, `call_function.rs`,
`prepare_function.rs`, and `stream_function.rs` for clarity.
> - Adjust import order in `ir_to_py/mod.rs` and `ir_to_rb/mod.rs` for
consistency.
>   - **Misc**:
> - Minor adjustments in `function_results.rs` to improve error handling
and code clarity.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 84920bd186187c576c70d255a6105ebb4b7d8d3b. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

antoniosarosi committed:
Fix devcontainer (#2063)

<!-- ELLIPSIS_HIDDEN -->



> [!IMPORTANT]
> Improve `README.md` in `integ-tests/ruby` with better formatting,
links, and development mode instructions.
> 
>   - **Documentation**:
> - Update `README.md` in `integ-tests/ruby` to improve formatting and
clarity.
> - Add links to `mise` documentation for Ruby installation management.
> - Include instructions for building in development mode using
`RB_SYS_CARGO_PROFILE`.
> - Remove redundant bash code block indicator for running specific
tests.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 15443947113009c090c03786e7de0d821ea48d22. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

antoniosarosi committed:
Add missing extensions to devcontainer (#2064)

<!-- ELLIPSIS_HIDDEN -->



> [!IMPORTANT]
> Remove trailing whitespace in `test_env_vars.rb`.
> 
>   - **Whitespace Removal**:
> - Remove trailing whitespace in `test_env_vars.rb` at lines 12, 15,
25, 30, 54, 70, and 97.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 906f43e4ee05f8629803a958aa6a5b6f2bba48c9. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

hellovai committed:
fix VSCode Proxy Handling due to v2 -> v3 for npm:http-proxy-middleware (#2065)

<!-- ELLIPSIS_HIDDEN -->



> [!IMPORTANT]
> Improves proxy handling in `extension.ts` for VSCode extension to
support `http-proxy-middleware` v3, including path rewriting, dynamic
target selection, and enhanced error handling.
> 
>   - **Proxy Handling**:
> - Updated `pathRewrite` to clear paths for image GET requests and
remove trailing slashes in `extension.ts`.
> - Enhanced `router` to dynamically choose target based on
`baml-original-url` header, ensuring no double-prefixing.
> - Improved error handling in `error` event to robustly report errors
and handle socket destruction.
>   - **Misc**:
>     - Added CORS header in `proxyRes` event to allow all origins.
> - Added logging for proxy requests and responses for better debugging.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 31ec83d9ce9bf03cfa35bf7c63acd365a1b525d6. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

antoniosarosi committed:
Add `cargo clippy` to CI (#2067)

<!-- ELLIPSIS_HIDDEN -->


> [!IMPORTANT]
> Add `cargo clippy` to CI and fix clippy warnings across the codebase
to enforce Rust linting rules.
> 
>   - **CI Integration**:
> - Add `cargo clippy` to `.github/workflows/primary.yml` for Rust
linting.
>     - Introduce steps for standard and WASM targets.
>   - **Code Changes**:
> - Fix clippy warnings by removing unnecessary references in
`attribute.rs`, `expression.rs`, and `mod.rs`.
> - Update error handling in `to_baml_arg.rs` to use `ArgCoerceError`
instead of `()`.
> - Simplify error handling in `repr.rs` by removing redundant
`or_else`.
>     - Correct scope usage in `expr_fns.rs` to avoid mutable borrow.
>     - Adjust logging and serialization in `logger.rs` and `lib.rs`.
>     - Update `Cargo.toml` files to allow specific clippy lints.
>   - **Miscellaneous**:
>     - Fix typos and improve code readability in various files.
> - Update test expectations in `input-output.test.ts` for consistency.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for 91dd423355b8d1544f93c2881d9ba45f12c16a47. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>


<!-- ELLIPSIS_HIDDEN -->

antoniosarosi committed:
Add `clippy` and `rustfmt` with import order configs to vscode on save (#2069)

<!-- ELLIPSIS_HIDDEN -->



> [!IMPORTANT]
> Add `clippy` and `rustfmt` configurations to VSCode workspace settings
for Rust projects.
> 
>   - **VSCode Workspace Configuration**:
> - Add `clippy` as the check command in `rust-analyzer` settings in
`root-wasm32.code-workspace` and `root.code-workspace`.
> - Configure `rustfmt` to use `imports_granularity=Crate` and
`group_imports=StdExternalCrate` in both workspace files.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis"
src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=BoundaryML%2Fbaml&utm_source=github&utm_medium=referral)<sup>
for ab077829bcff7a31d7e47f89daf9ac8fbfff1fad. You can
[customize](https://app.ellipsis.dev/BoundaryML/settings/summaries) this
summary. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

egol committed:
Merge branch 'egor/zed-extension' into canary

egol committed:
regenerate cargo lock

egol committed:
run cargo format

egol committed:
build language server deps in for lint

egol committed:
clippy fixes

egol committed:
fix formatter
